src/combinatorics/simple_graph/matching.lean

```diff
@@ -1,69 +1,82 @@
 -/
-import combinatorics.simple_graph.basic
+import combinatorics.simple_graph.subgraph
+import combinatorics.simple_graph.degree_sum
 
 /-!
 # Matchings
 
+The idea of a matching refers to the pairing of adjacent vertices. In this approach we're
+representing a matching as a subgraph that contains all vertices of a graph.
 
 ## Main definitions
 
-* a `matching` on a simple graph is a subset of its edge set such that
-   no two edges share an endpoint.
+* A *matching* `M` on a simple graph `G` is a subgraph of `G`, such that every vertex of `G` is a
+  vertex of `M` and no two edges of `M` share an endpoint.
 
-* a `perfect_matching` on a simple graph is a matching in which every
-   vertex belongs to an edge.
+* A *perfect matching* on a simple graph is a matching in which every vertex forms an edge.
 
-TODO:
-  - Lemma stating that the existence of a perfect matching on `G` implies that
-    the cardinality of `V` is even (assuming it's finite)
-  - Hall's Marriage Theorem (see combinatorics.hall)
-  - Tutte's Theorem
-  - consider coercions instead of type definition for `matching`:
-    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532935457
-  - consider expressing `matching_verts` as union:
-    https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532906131
+## Todo
 
-TODO: Tutte and Hall require a definition of subgraphs.
--/
-open finset
-universe u
+* Lemma stating that the existence of a perfect matching on `G` implies that
+  the cardinality of `V` is even (assuming it's finite)
 
-namespace simple_graph
-variables {V : Type u} (G : simple_graph V)
+* Hall's Marriage Theorem (see combinatorics.hall)
 
-/--
-A matching on `G` is a subset of its edges such that no two edges share a vertex.
+* Tutte's Theorem
+
+* https://github.com/leanprover-community/mathlib/pull/5156#discussion_r532906131
 -/
-structure matching :=
-(edges : set (sym2 V))
-(sub_edges : edges ⊆ G.edge_set)
-(disjoint : ∀ (x y ∈ edges) (v : V), v ∈ x ∧ v ∈ y → x = y)
 
-instance : inhabited (matching G) :=
-⟨⟨∅, set.empty_subset _, λ _ _ hx, false.elim (set.not_mem_empty _ hx)⟩⟩
+universe u
 
-variables {G}
+namespace simple_graph
+variables {V : Type u} {G : simple_graph V} (M : subgraph G)
 
 /--
-`M.support` is the set of vertices of `G` that are
-contained in some edge of matching `M`
+The subgraph `M` of `G` is a matching on `G` if every vertex only forms at most one edge of `M`.
 -/
-def matching.support (M : G.matching) : set V :=
-{v : V | ∃ x ∈ M.edges, v ∈ x}
+def is_matching : Prop := ∀ (v w w': V), M.adj v w → M.adj v w' → w = w'
 
 /--
-A perfect matching `M` on graph `G` is a matching such that
-  every vertex is contained in an edge of `M`.
+`M` is a perfect matching on `G` if it's a matching on `G` and if every vertex forms an edge of `M`.
 -/
-def matching.is_perfect (M : G.matching) : Prop :=
-M.support = set.univ
+def is_perfect : Prop := is_matching M ∧ M.support = set.univ
```

Comment:
This should be `is_perfect_matching` now.

Given the comment about needing `M.verts = M.support`, let's change this to `M.is_spanning`.
```suggestion
def is_perfect_matching : Prop := is_matching M ∧ M.is_spanning
```